processor/src/host/advice/mod.rs

```diff
@@ -143,9 +150,48 @@ impl AdviceProvider {
     // ADVICE MAP
     // --------------------------------------------------------------------------------------------
 
-    /// Returns a reference to the value(s) associated with the specified key in the advice map.
-    pub fn get_mapped_values(&self, key: &Word) -> Result<&[Felt], AdviceError> {
-        self.map.get(key).ok_or(AdviceError::MapKeyNotFound { key: *key })
+    /// Returns a reference to the value(s) associated with the specified key in the advice map,
+    /// searching secondary storage like `MastForest`s if necessary.
+    ///
+    /// # Details
+    /// This lookup is performed in two stages as a performance optimization:
+    /// 1. The primary advice map is checked first. If the key is present, the value is returned
+    ///    immediately.
+    /// 2. If the key is not found, the available [`MastForest`]s are searched. If a forest
+    ///    containing the key is located, its [`AdviceMap`] is merged into this provider's primary
+    ///    map, and the value can then be retrieved.
+    ///
+    /// This lazy-loading approach avoids the overhead of merging all forest maps upfront.
+    ///
+    /// # Errors
+    /// Returns an error if the key is not found in the primary map or in any of the available
+    /// `MastForest`s.
+    pub fn get_mapped_values(&mut self, key: &Word) -> Result<&[Felt], AdviceError> {
+        // Hot Path: Perform a single lookup.
+        if let Some(values) = self.map.get(key) {
+            // --- UNSAFE ---
+            // SAFETY: This is safe because this branch terminates immediately with a return.
+            // No mutation of `self` happens, so the returned reference cannot be
+            // invalidated by later operations in this function. We are manually
+            // telling the compiler that the aliasing it fears does not actually occur.
+            unsafe {
```

Comment:
The alternative is for `AdviceMap::get` to return `Option<Arc[Felt]>` which would always clone the pointer.

Reply:
Let's not have unsafe code here (unless we really benchmark that this ends up speeding up an important class of programs). A slightly less efficient but probably very fine option is:

```rs
    pub fn get_mapped_values(&mut self, key: &Word) -> Result<&[Felt], AdviceError> {
        if self.map.contains_key(key) {
            return Ok(self.map.get(key).unwrap());
        }

        if let Some(mast_forest) =
            self.pending_forests.pop_if(|mf| mf.advice_map().contains_key(key))
        {
            self.merge_advice_map(mast_forest.advice_map())?;
            Ok(self.map.get(key).unwrap())
        } else {
            Err(AdviceError::MapKeyNotFound { key: *key })
        }
    }
```